### DISCUSSION TITLE
Historical option data for non-expired contracts

### DISCUSSION THREAD

**Discussion by Khrizz9000** [erdewit/ib_insync#446] [Q&A]
Hi all,

I would like to get historical data (ideally hourly or less) for non-expired option contracts. TWS API should provide that data (source: https://www.tradersinsight.news/ibkr-quant-news/historical-options-futures-data-using-tws-api/). I have not found a way to get that data using the beautiful ib_insync lib. Is there a way to do that? If so, how?

Thx in advance

**Comment by m12t** [comment]
Have you been able to request historical data for other underlyings such as SPX or AAPL Stock? I don't know exactly what your familiarity is with the library and IBKR. If so, you should be able to simply create the contract object for the underlying you wish (the unexpired option contract) and the request historical data on it. I've recreated a MRE below that does just this getting a 1 day duration of 1-minute data:
```py
contract = Option(
    symbol='SPX', lastTradeDateOrContractMonth='20220304',
    strike=4400, right='C', exchange='CBOE',
    multiplier=100, currency='USD')

bars = ib.reqHistoricalData(
            contract=contract, endDateTime='20220228 16:00:00 EST', durationStr='1 D', barSizeSetting='1 min',
            whatToShow='TRADES', useRTH=True)
```

After converting the data to a pandas dataframe via ```util.df(bars)``` for convenience, you can validate that the data are as expected:

```
                   date  open  high   low  close  volume  average  barCount
610 2022-02-28 08:30:00  33.7  33.7  33.7   33.7       3     33.7         1
611 2022-02-28 08:31:00  34.5  34.5  34.5   34.5      10     34.5         1
612 2022-02-28 08:32:00  34.5  34.5  34.5   34.5       0     34.5         0
613 2022-02-28 08:33:00  34.5  34.5  34.5   34.5       0     34.5         0
614 2022-02-28 08:34:00  32.8  32.8  32.8   32.8       1     32.8         1
..                  ...   ...   ...   ...    ...     ...      ...       ...
995 2022-02-28 14:55:00  34.2  34.2  34.2   34.2       0     34.2         0
996 2022-02-28 14:56:00  34.2  34.2  34.2   34.2       0     34.2         0
997 2022-02-28 14:57:00  34.2  34.2  34.2   34.2       0     34.2         0
998 2022-02-28 14:58:00  34.2  34.2  34.2   34.2       0     34.2         0
999 2022-02-28 14:59:00  34.2  34.2  34.2   34.2       0     34.2         0
```

**Reply by EugeniuZ** [reply]
What is the meaning of `average` and `barCount` columns? Is it the VWAP and number of trades?

**Reply by erdewit** [reply]
> Is it the VWAP and number of trades?

Yes it is.

**Reply by August1328** [reply]
Hi all,

are we sure that the values in the `average` column are the VWAP?

I have been using `ib.reqHistoricalData()` for over a year and back then in my script I decided to drop the `average` column - unfortunately I didn´t write down or remember why.

I came upon this discussion here a few days ago and thought it would be great to get the VWAP with the historical data.

Today I wrote a new script and tested `ib.reqHistoricalData()` with a few stock symbols, but the values in `average ` are not the same as the VWAP on the chart or the values from `ticker.vwap` when calling `genericTick ` 233.

Any opinion on this?

Thanks
